iceberg/src/main/scala/org/apache/iceberg/spark/source/GpuIcebergPartitionReader.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright (c) 2025, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.iceberg.spark.source
+
+import java.nio.ByteBuffer
+import java.util.{Map => JMap}
+
+import scala.collection.JavaConverters._
+
+import com.nvidia.spark.rapids.GpuMetric
+import com.nvidia.spark.rapids.iceberg.parquet.{GpuCoalescingIcebergParquetReader, GpuIcebergParquetReader, GpuIcebergParquetReaderConf, GpuMultiThreadIcebergParquetReader, GpuSingleThreadIcebergParquetReader, IcebergPartitionedFile, MultiFile, MultiThread, SingleFile, ThreadConf}
+import org.apache.avro.generic.GenericData
+import org.apache.avro.util.Utf8
+import org.apache.iceberg.{FileFormat, FileScanTask, MetadataColumns, Partitioning, ScanTask, ScanTaskGroup, Schema, StructLike, Table, TableProperties}
+import org.apache.iceberg.encryption.EncryptedFiles
+import org.apache.iceberg.mapping.NameMappingParser
+import org.apache.iceberg.types.{Type, Types}
+import org.apache.iceberg.types.Type.TypeID
+import org.apache.iceberg.util.{ByteBuffers, PartitionUtil}
+
+import org.apache.spark.sql.catalyst.expressions.GenericInternalRow
+import org.apache.spark.sql.connector.read.PartitionReader
+import org.apache.spark.sql.types.Decimal
+import org.apache.spark.sql.vectorized.ColumnarBatch
+import org.apache.spark.unsafe.types.UTF8String
+
+
+class GpuIcebergPartitionReader(private val task: GpuSparkInputPartition,
+    private val threadConf: ThreadConf,
+    private val metrics: Map[String, GpuMetric],
+) extends PartitionReader[ColumnarBatch] {
+  private lazy val table = task.cpuPartition.table()
+  private lazy val fileIO = table.io()
+  private lazy val conf = newConf()
+  private lazy val (_, tasks) = collectFiles()
+  private lazy val reader: GpuIcebergParquetReader = createDataFileParquetReader()
+
+  private var inited = false
```

Comment:
nit: Just to make thing more clear, especially with scala not having a formal constructor, could we declare this before reader?